Eradicate the ability to create new sstables with numerical sstable generation

bhalevy opened the pull request:
Remove support for generating numerical sstable generation for new sstables.
Loading such sstables is still supported but new sstables are always created with a uuid generation.
This is possible since:
* All live versions (since 5.4 / f014ccf36962135889a84cff15b0478d711b2306) now support uuid sstable generations.
* The `uuid_sstable_identifiers_enabled` config option (that is unused from version 2025.2 / 6da758d74c83916dc1f2eff99084062e77b89914) controls only the use of uuid generations when creating new sstables. SSTables with uuid generations should still be properly loaded by older versions, even if `uuid_sstable_identifiers_enabled` is set to `false`.

Fixes #24248

* Enhancement, no backport needed

bhalevy commented:
Cc @regevran

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17691/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17691/)
:x: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17691/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/database_test
:small_blue_diamond: boost/sstable_compaction_test
:small_blue_diamond: boost/sstable_datafile_test
:small_blue_diamond: boost/sstable_directory_test
:small_blue_diamond: boost/sstable_move_test
:small_blue_diamond: boost/sstable_resharding_test
:small_blue_diamond: boost/sstable_test
#### Failed Tests (3/272):
- [test_sstable_move](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17691/testReport/junit/boost/sstable_move_test/test_sstable_move) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstable_move)
- [test_sstable_move_idempotent](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17691/testReport/junit/boost/sstable_move_test/test_sstable_move_idempotent) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstable_move_idempotent)
- [test_sstable_move_replay](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17691/testReport/junit/boost/sstable_move_test/test_sstable_move_replay) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sstable_move_replay)
#### Build Details:
- Duration: 3 hr 25 min
- Builder: spider8.cloudius-systems.com

bhalevy commented:
> > Remove support for generating numerical sstable generation for new sstables. Loading such sstables is still supported but new sstables are always created with a uuid generation. This is possible since all live versions now support uuid sstable generations.
> 
> Please name the version that introduced mandatory uuid generations and its release date.

f014ccf36962135889a84cff15b0478d711b2306 introduced uuid generations as on by default and it's present since version 5.4 (2023-04-12, but [announced](https://www.scylladb.com/2023/12/11/scylladb-open-source-5-4/) only on December '23).

They became mandatory, in the sense that the config option cannot be disabled only recently in 6da758d74c83916dc1f2eff99084062e77b89914 which will be released only in 2025.2.0
The question is when did the deprecation clock start ticking?

bhalevy commented:
In [67af4b2](https://github.com/scylladb/scylladb/commit/67af4b2c1fa7e8ca1a2a1d3b0e0b261277b70365):
- Adjusted sstable_move_test

denesb commented on replica/distributed_loader.cc:
```diff
@@ -198,9 +198,8 @@ distributed_loader::process_upload_dir(distributed<replica::database>& db, shard
-            auto generation = sharded_gen.invoke_on(shard, [uuid_sstable_identifiers] (auto& gen) {
-                return gen(sstables::uuid_identifiers{uuid_sstable_identifiers});
+            auto generation = sharded_gen.invoke_on(shard, [] (auto& gen) {
+                return gen();
```
Is this hunk related?

bhalevy replied:
Hmm, looks like a bad squash.
Should be in "sstables: sstable_generation: get rid of uuid_identifiers bool class"

denesb reviewed (approved):
Now that we don't need to track highest generation anymore, can we make generation generator completely stateless and remove the generator from make sstable factory method parameters?

bhalevy commented:
In [d5bcc43](https://github.com/scylladb/scylladb/commit/d5bcc4304a97ab6fe39789021f24d132b60452a4):
- Refactored feature_service: never disable UUID_SSTABLE_IDENTIFIERS
  - Split out sstables_manager: drop uuid_sstable_identifiers
- replica: distributed_loader: process_upload_dir: use the table sstable_generation_generator
- streaming: stream_blob: use the table sstable_generation_generator

denesb commented:
> > Follow-up idea: now that we don't need to track highest generation anymore, can we make generation generator completely stateless and remove the generator from make sstable factory method parameters?
> 
> I'm not sure which factory you mean. Can you please be more specific?



> > Follow-up idea: now that we don't need to track highest generation anymore, can we make generation generator completely stateless and remove the generator from make sstable factory method parameters?
> 
> I'm not sure which factory you mean. Can you please be more specific?

`replica::table::_sstable_generation_generator`. Seems like it is completely stateless now, so we can get rid of it (of the entire object) and replace with a function `sstables::new_generation()`.

bhalevy commented:
> > > Follow-up idea: now that we don't need to track highest generation anymore, can we make generation generator completely stateless and remove the generator from make sstable factory method parameters?
> > 
> > 
> > I'm not sure which factory you mean. Can you please be more specific?
> 
> > > Follow-up idea: now that we don't need to track highest generation anymore, can we make generation generator completely stateless and remove the generator from make sstable factory method parameters?
> > 
> > 
> > I'm not sure which factory you mean. Can you please be more specific?
> 
> `replica::table::_sstable_generation_generator`. Seems like it is completely stateless now, so we can get rid of it (of the entire object) and replace with a function `sstables::new_generation()`.

The generation does encode the shard id, so we need sharded objects, hence having a generator per table shard is useful.

avikivity commented:
> > > Remove support for generating numerical sstable generation for new sstables. Loading such sstables is still supported but new sstables are always created with a uuid generation. This is possible since all live versions now support uuid sstable generations.
> > 
> > 
> > Please name the version that introduced mandatory uuid generations and its release date.
> 
> [f014ccf](https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306) introduced uuid generations as on by default and it's present since version 5.4 (2023-04-12, but [announced](https://www.scylladb.com/2023/12/11/scylladb-open-source-5-4/) only on December '23).
> 
> They became mandatory, in the sense that the config option cannot be disabled only recently in [6da758d](https://github.com/scylladb/scylladb/commit/6da758d74c83916dc1f2eff99084062e77b89914) which will be released only in 2025.2.0 The question is when did the deprecation clock start ticking?

We can't remove them so soon.

If we want to allow migration from 2023.1 to 2026.1, we can't merge this.

mykaul commented:
> If we want to allow migration from 2023.1 to 2026.1, we can't merge this

I hope we have no such plans.

bhalevy commented:
[3b2edc0](https://github.com/scylladb/scylladb/commit/3b2edc017c3407c7955554409130fc35a5cccd25): rebased

avikivity commented:
> This is possible since all live versions now support uuid sstable generations.


Which patch introduced it?

It can still be disabled in all versions, will they recognize it?

This information should be in the cover letter.

bhalevy commented:
> > This is possible since all live versions now support uuid sstable generations.
> 
> Which patch introduced it?

https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306 / version 5.4

> 
> It can still be disabled in all versions, will they recognize it?

Yes, sstables with uuid generations should be loaded normally with older versions, even when the config option is disabled.
 
> This information should be in the cover letter.

@avikivity I've updated the cover letter.

avikivity commented:
> > > This is possible since all live versions now support uuid sstable generations.
> > 
> > 
> > Which patch introduced it?
> 
> [f014ccf](https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306) / version 5.4
> 
> > It can still be disabled in all versions, will they recognize it?
> 
> Yes, sstables with uuid generations should be loaded normally with older versions, even when the config option is disabled.

Then "eradicate" is an exaggeration.

Is this tested?

avikivity commented:
> > > This is possible since all live versions now support uuid sstable generations.
> > 
> > 
> > Which patch introduced it?
> 
> [f014ccf](https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306) / version 5.4
> 

That's 2024.1.

If we prevent numerical generations now, it will prevent 2024.1 -> 2026.1 upgrades.

Suggest to delay to 2026.2.

avikivity commented:
> > > > > This is possible since all live versions now support uuid sstable generations.
> > > > 
> > > > 
> > > > Which patch introduced it?
> > > 
> > > 
> > > [f014ccf](https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306) / version 5.4
> > 
> > 
> > That's 2024.1.
> > If we prevent numerical generations now, it will prevent 2024.1 -> 2026.1 upgrades.
> 
> Why? 2024.1 supports uuid generations (with [f014ccf](https://github.com/scylladb/scylladb/commit/f014ccf36962135889a84cff15b0478d711b2306))

Right, miscalculated.

> 
> > Suggest to delay to 2026.2.
> 
> I'm fine with that, but it seems over-conservative to me.
> 
> > > Yes, sstables with uuid generations should be loaded normally with older versions, even when the config option is disabled.
> > 
> > 
> > Then "eradicate" is an exaggeration.
> > Is this tested?
> 
> There is no test for it. I can test it manually to make sure that my understanding of the code is indeed correct.

Yes please.

bhalevy commented:
@avikivity so once the feature is enabled, there is no automatic way back.

I got this with 2025.1:
```
ERROR 2025-08-03 20:04:29,086 [shard 0:main] init - Startup failed: gms::unsupported_feature_exception (Feature 'UUID_SSTABLE_IDENTIFIERS' was previously enabled in the cluster but its support is disabled by this node. Set the corresponding configuration option to enable the support for the feature.)
```

Using a single node test that just flips the configuration value:
```python
@pytest.mark.asyncio
async def test_sstable_generation_type(manager: ManagerClient):
    """
    Test single node enabling and disabling of uuid_sstable_identifiers_enabled
    """
    node = await manager.server_add(config={'uuid_sstable_identifiers_enabled': False})

    ks = await create_new_test_keyspace(manager.get_cql(), "with replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
    t = unique_name()
    await manager.get_cql().run_async(f"CREATE TABLE {ks}.{t} (k int PRIMARY KEY, v int) WITH compaction = {{'class': 'NullCompactionStrategy'}}")

    stmt = manager.get_cql().prepare(f"INSERT INTO {ks}.{t} (k, v) VALUES (?, ?)")

    async def restart_server(uuid_sstable_identifiers_enabled: bool):
        await manager.server_stop_gracefully(node.server_id)
        await manager.server_update_config(node.server_id, key='uuid_sstable_identifiers_enabled', value=uuid_sstable_identifiers_enabled)
        await manager.server_start(node.server_id)
        await wait_for_cql_and_get_hosts(manager.get_cql(), [node], time.time() + 60)

    async def validate_count(expected_count):
        res = await manager.get_cql().run_async(f"SELECT count(*) FROM {ks}.{t}")
        assert res[0][0] == expected_count, f"Expected {expected_count} rows, but got {res[0][0]}"

    await manager.get_cql().run_async(stmt, (0, 0))
    await manager.api.client.post_json("/storage_service/flush", host=node.ip_addr)

    await restart_server(False)
    await validate_count(1)

    await manager.get_cql().run_async(stmt, (1, 1))
    await manager.api.client.post_json("/storage_service/flush", host=node.ip_addr)

    await restart_server(True)
    await validate_count(2)

    # With this the server fails to restart due to attempt to disable a cluster feature
    await restart_server(False)
    await validate_count(2)
```

avikivity commented:
> Manually tested also on 2024.1 The conclusion is that the option can be safely enabled on old releases before upgrading. 

What if someone forgets to do that?


> Restarting the nodes would load sstables both with numerical or with uuid generations. Running major compaction (or upgrade sstables) would rewrite the sstables with uuid generations.

Why is that needed? We aren't removing the ability to read numberical generations.

raphaelsc commented:
On Mon, Aug 4, 2025, 7:25 AM Benny Halevy ***@***.***> wrote:

> *bhalevy* left a comment (scylladb/scylladb#24512)
> <https://github.com/scylladb/scylladb/pull/24512#issuecomment-3150217065>
>
> Manually tested also on 2024.1 The conclusion is that the option can be
> safely enabled on old releases before upgrading.
>
> What if someone forgets to do that?
>
> Then upgrade rollback would fail Feature 'UUID_SSTABLE_IDENTIFIERS' was
> previously enabled in the cluster but its support is disabled by this node.
>
Maybe we need to go another step of deprecation dance before we can erase
integer generations? We need to make sure inactive tables have their
sstables rewritten. Otherwise such a table becoming active again couldn't
have their sstables with integer gen loaded



Restarting the nodes would load sstables both with numerical or with uuid
> generations. Running major compaction (or upgrade sstables) would rewrite
> the sstables with uuid generations.
>
> Why is that needed? We aren't removing the ability to read numberical
> generations.
>
> Right, that's not a requirement.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/24512#issuecomment-3150217065>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA45BMG2GMYFIB4Y5GE33L47LJAVCNFSM6AAAAAB7LHG3DKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNJQGIYTOMBWGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

bhalevy commented:
> On Mon, Aug 4, 2025, 7:25 AM Benny Halevy ***@***.***> wrote: *bhalevy* left a comment ([scylladb/scylladb#24512](https://github.com/scylladb/scylladb/pull/24512)) <[#24512 (comment)](https://github.com/scylladb/scylladb/pull/24512#issuecomment-3150217065)> Manually tested also on 2024.1 The conclusion is that the option can be safely enabled on old releases before upgrading. What if someone forgets to do that? Then upgrade rollback would fail Feature 'UUID_SSTABLE_IDENTIFIERS' was previously enabled in the cluster but its support is disabled by this node.
> Maybe we need to go another step of deprecation dance before we can erase integer generations? We need to make sure inactive tables have their sstables rewritten. Otherwise such a table becoming active again couldn't have their sstables with integer gen loaded

Loading is still supported.
It's only the generation of new sstables with integer generations that is eradicated.

raphaelsc commented:
> > On Mon, Aug 4, 2025, 7:25 AM Benny Halevy ***@***.***> wrote: *bhalevy* left a comment ([scylladb/scylladb#24512](https://github.com/scylladb/scylladb/pull/24512)) <[#24512 (comment)](https://github.com/scylladb/scylladb/pull/24512#issuecomment-3150217065)> Manually tested also on 2024.1 The conclusion is that the option can be safely enabled on old releases before upgrading. What if someone forgets to do that? Then upgrade rollback would fail Feature 'UUID_SSTABLE_IDENTIFIERS' was previously enabled in the cluster but its support is disabled by this node.
> > Maybe we need to go another step of deprecation dance before we can erase integer generations? We need to make sure inactive tables have their sstables rewritten. Otherwise such a table becoming active again couldn't have their sstables with integer gen loaded
> 
> Loading is still supported.
> It's only the generation of new sstables with integer generations that is eradicated.

Maybe we should postpone this change until we know all supported versions work with uuid generation? Similar to how we deprecate a feature flag only after we consider all supported installations have it enabled.

bhalevy commented:
> > > > Manually tested also on 2024.1 The conclusion is that the option can be safely enabled on old releases before upgrading.
> > > 
> > > 
> > > What if someone forgets to do that?
> > 
> > 
> > Then upgrade rollback would fail `Feature 'UUID_SSTABLE_IDENTIFIERS' was previously enabled in the cluster but its support is disabled by this node.`
> 
> Then they'd be stuck, no?
> 
> Or can they enable it in scylla.yaml and proceed?

It's the latter case, like you said, they can enable uuid sstable generations in scylla.yaml and proceed with the rollback.

avikivity commented:
> > > > > Manually tested also on 2024.1 The conclusion is that the option can be safely enabled on old releases before upgrading.
> > > > 
> > > > 
> > > > What if someone forgets to do that?
> > > 
> > > 
> > > Then upgrade rollback would fail `Feature 'UUID_SSTABLE_IDENTIFIERS' was previously enabled in the cluster but its support is disabled by this node.`
> > 
> > 
> > Then they'd be stuck, no?
> > Or can they enable it in scylla.yaml and proceed?
> 
> It's the latter case, like you said, they can enable uuid sstable generations in scylla.yaml and proceed with the rollback.

Ok. And it defaults to enabled, so likelihood of (disabled) && (need rollback) is low.

avikivity commented:
We'll lose the ability to import numerically generationed sstables. I suppose it's easy to provide a script to rename them.

Won't we have failing tests? I see gating passed, but won't we have `nodetool refresh` tests elsewhere that weren't updated to uuid generations?

bhalevy commented:
> We'll lose the ability to import numerically generationed sstables. I suppose it's easy to provide a script to rename them.

We don't lose the ability to load sstables with numerical generations, just the ability to generate new ones.
And the code that tracked the highest generation number that was used for that.
 
> Won't we have failing tests? I see gating passed, but won't we have `nodetool refresh` tests elsewhere that weren't updated to uuid generations?

Like I said, we can still load sstables with numerical generations.

bhalevy commented:
In [9c18c22](https://github.com/scylladb/scylladb/commit/9c18c229b34e760134af12c20d95c9e78b420d51):
- Added unit test verifying that sstables with mixed numerical and uuid generations can be loaded properly
- feature_service: add UUID_SSTABLE_IDENTIFIERS to supported_feature_set
  - making it always enabled using the official method

avikivity commented:
> > We'll lose the ability to import numerically generationed sstables. I suppose it's easy to provide a script to rename them.
> 
> We don't lose the ability to load sstables with numerical generations, just the ability to generate new ones. And the code that tracked the highest generation number that was used for that.

Then "eradicate" is an exaggeration.

> 
> > Won't we have failing tests? I see gating passed, but won't we have `nodetool refresh` tests elsewhere that weren't updated to uuid generations?
> 
> Like I said, we can still load sstables with numerical generations.

That's good. But the title is misleading.

bhalevy commented:
> > > We'll lose the ability to import numerically generationed sstables. I suppose it's easy to provide a script to rename them.
> > 
> > 
> > We don't lose the ability to load sstables with numerical generations, just the ability to generate new ones. And the code that tracked the highest generation number that was used for that.
> 
> Then "eradicate" is an exaggeration.
> 
> > > Won't we have failing tests? I see gating passed, but won't we have `nodetool refresh` tests elsewhere that weren't updated to uuid generations?
> > 
> > 
> > Like I said, we can still load sstables with numerical generations.
> 
> That's good. But the title is misleading.

Rephrased the title

avikivity commented on gms/feature_service.hh:
```diff
@@ -113,6 +113,7 @@ public:
     gms::feature secondary_indexes_on_static_columns { *this, "SECONDARY_INDEXES_ON_STATIC_COLUMNS"sv };
     gms::feature tablets { *this, "TABLETS"sv };
+    // UUID_SSTABLE_IDENTIFIERS are always enabled
     gms::feature uuid_sstable_identifiers { *this, "UUID_SSTABLE_IDENTIFIERS"sv };
```
It should be removed in the same patch, I don't know how the code reacts to having it both in a feature and in the grandfathered set.

bhalevy commented:
> Did CI not run here?

I removed the `conflicts` label and indeed CI dis not run since.

@yaronkaikov ism't it supposed to be trigerrred in this case?

yaronkaikov commented:
> > Did CI not run here?
> 
> I removed the `conflicts` label and indeed CI dis not run since.
> 
> @yaronkaikov ism't it supposed to be trigerrred in this case?

removing the label only will not trigger the CI, you need also to push changes to branch (or empty commit)

bhalevy commented:
> > > Did CI not run here?
> > 
> > 
> > I removed the `conflicts` label and indeed CI dis not run since.
> > @yaronkaikov ism't it supposed to be trigerrred in this case?
> 
> removing the label only will not trigger the CI, you need also to push changes to branch (or empty commit)

@yaronkaikov I'm not even sure when the conflicts label was even added, and so I pushed a new version and only them dropped the label.  It should be autonatically removed when the PR branch is force pushed, and auto-added only on new merge conflicts.

bhalevy commented:
CI hit https://github.com/scylladb/scylladb/issues/25427

bhalevy commented:
@avikivity please re-review.
I believe the series is ready now.
